src/libraries/System.Private.CoreLib/src/System/IO/BufferedFileStreamStrategy.cs

```diff
@@ -0,0 +1,1079 @@
+                // when users have a race condition in their code (i.e. they call
+                // FileStream.Close when calling another method on FileStream like Read).
+
+                _writePos = 0; // WriteByte hot path relies on this
+            }
+        }
+
+        internal override void DisposeInternal(bool disposing) => Dispose(disposing);
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing && !_strategy.IsClosed)
+                {
+                    try
+                    {
+                        Flush();
+                    }
+                    finally
+                    {
+                        _strategy.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                // Don't set the buffer to null, to avoid a NullReferenceException
+                // when users have a race condition in their code (i.e. they call
+                // FileStream.Close when calling another method on FileStream like Read).
+
+                // Call base.Dispose(bool) to cleanup async IO resources
+                base.Dispose(disposing);
+
+                _writePos = 0;
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            AssertBufferArguments(buffer, offset, count);
+
+            return ReadSpan(new Span<byte>(buffer, offset, count), new ArraySegment<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> destination)
+        {
+            EnsureNotClosed();
+
+            return ReadSpan(destination, default);
+        }
+
+        private int ReadSpan(Span<byte> destination, ArraySegment<byte> arraySegment)
+        {
+            Debug.Assert((_readPos == 0 && _readLen == 0 && _writePos >= 0) || (_writePos == 0 && _readPos <= _readLen),
+                "We're either reading or writing, but not both.");
+
+            bool isBlocked = false;
+            int n = _readLen - _readPos;
+            // if the read buffer is empty, read into either user's array or our
+            // buffer, depending on number of bytes user asked for and buffer size.
+            if (n == 0)
+            {
+                EnsureCanRead();
+
+                if (_writePos > 0)
+                {
+                    FlushWrite();
+                }
+
+                if (!_strategy.CanSeek || (destination.Length >= _bufferSize))
+                {
+                    // For async file stream strategies the call to Read(Span) is translated to Stream.Read(Span),
+                    // which rents an array from the pool, copies the data, and then calls Read(Array). This is expensive!
+                    // To avoid that (and code duplication), the Read(Array) method passes ArraySegment to this method
+                    // which allows for calling Strategy.Read(Array) instead of Strategy.Read(Span).
```

Comment:
> Is there a way to avoid that penalty without special-casing here?

I am afraid not for the reasons described below.

> Why AsyncFileStreamStrategy doesn't override Read(Span) in order to avoid renting an array from the pool, copying the data, and then call Read(Array)?

`Span` is stack-only (https://adamsitnik.com/Span/#stack-only). Stacks are not shared across multiple threads, so a single Stack is accessed by one thread at the same time.
`AsyncFileStreamStrategy` must perform all operations in an async way and after an await a different thread can be continuing the execution, which would mean that given `Span` points to invalid memory. This is why we have to perform this copy. This is also why we have been doing it so far (5.0 code  below):

https://github.com/dotnet/runtime/blob/55e7712b5328ce37042210ffa37dc6adbf645bf5/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L309-L329